### DISCUSSION TITLE
Transactions / undoing a script / bulk undo, Q&A + some musings

### DISCUSSION THREAD

**Discussion by allgreed** [GothenburgBitFactory/taskwarrior#3132] [General]
Hey, I've been using taskwarrior for quite a while now and was wondering about bulk undo.

### Usecase 1:
```
t~ >_ t add a
Created task 134.
~ >_ t add b
Created task 135.
~ >_ t add v
Created task 136.
~ >_ t done 134 135 136
  - End will be set to '10.07'.
  - Status will be changed from 'pending' to 'completed'.
Complete task 134 'a'? (yes/no/all/quit) a
Completed task 134 'a'.
Completed task 135 'b'.
Completed task 136 'v'.
Completed 3 tasks.
~ >_ t undo

The last modification was made 2023-07-10

             Prior Values                          Current Values                      
description  v                                     v                                   
entry        10.07                                 10.07
modified     10.07                                 10.07                               
status       pending                               completed                           
uuid         12eb844c-21f1-49aa-8e55-840d7ea1d0e3  12eb844c-21f1-49aa-8e55-840d7ea1d0e3
end                                                10.07                               

The undo command is not reversible.  Are you sure you want to revert to the previous state? (yes/no)
```

So... here I'm running 1 command - there is one intent, yet I can only undo the last effect. And it would actually make sense to be able to undo all effects or just parts of it (think `git add -p`, but as an interface to undo).

### Usecase 2:
I have some scripts that create recurring tasks for me (semi-manual approach works best for me so far, as my needs are semi-regular in a way that's non-trivial to encode deterministically lol, but that's a story for another time). And scripts being code sometimes contain bugs, especially at the development stage. Effectively the same problem occurs as above - if my scripts does N things I have to undo them on a one-by-one basis, even tough I wouldn't have a problem with doing something like:
```
task _transaction_start
./do-shit.sh param1 param2
task _transaction_end
```
therefore marking the entire script as a single undoable intent (and ofc eventually moving the transaction boundary inside the script itself). I could then iterate really fast, since redo is basically free.

Now, I'm not super sure, I'd like to request a feature. It doesn't bother me _that_ much. But I'm wondering:
- does this strike a chord with someone else?
- how do you deal with this problem?

**Comment by djmitche** [comment]
The new backend tracks undo quite a bit differently. Basically it puts "markers" into the series of operations performed on the database, and `task undo` just undoes operations until it hits a marker. So in principle, `t done x,y,z` could just put a marker at the beginning of the command and then mark all three tasks as done. In fact, I suspect that's what it will do in the development branch.

Regarding the script use-case, I have a few thoughts:
 - Maybe it makes sense to make the undo behavior of the command-line tool configurable, so for example `task undo` would undo to the last time you ran `task undo-marker`.
 - I'd like to get TaskWarrior to the point where scripts run with their _own_ replica, and use programmatic APIs rather than running `task` as a separate process. Then `task sync` is used to combine the results of the scripts with the replica you use day-to-day. If that's the case, then the script could control its undo behavior directly (the programmatic API provides direct control of undo markers) and during development you could just disable sync until you're sure the script is working correctly.

Regarding a feature request -- to set your expectations, there's not a lot of developer bandwidth to work on things right now. So, it's quite likely a feature request would go un-answered for quite a while. That doesn't mean you shouldn't file one :)

**Reply by allgreed** [reply]
Wow, that's splendid to hear, so the core functionality is already there o.0

Ad scripts operating on their own replica: that's a very interesting approach - what's preventing it right now? I haven't got around to `sync` yet, but what you've written has piqued my curiosity.

**Reply by djmitche** [reply]
Nothing prevents that right now, but it's not the recommended setup. And, the current sync support is pretty buggy and needs a lot of TLC to keep it running.

**Comment by allgreed** [comment]
Out of curiosity - how has Taskwarrior 3 / Taskchampion affected this? I'm personally still not using v3 due to performance issues, but would love to get into it once I switch ;)

**Reply by djmitche** [reply]
Affected undo specifically? I addressed that [in the comment above](https://github.com/GothenburgBitFactory/taskwarrior/discussions/3132#discussioncomment-6409804). Basically, `task undo` will undo whatever the last `task` invocation that modified anything did, regardless of how many tasks it touched.
